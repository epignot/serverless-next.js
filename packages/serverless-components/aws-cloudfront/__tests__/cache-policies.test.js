const { createComponent, assertHasCacheBehavior } = require("../test-utils");

const {
  mockCreateDistribution,
  mockCreateDistributionPromise,
  mockCreateCachePolicyPromise,
  mockCreateCachePolicy
} = require("aws-sdk");

jest.mock("aws-sdk", () => require("../__mocks__/aws-sdk.mock"));

describe("Cache Policies", () => {
  let component;
  let testCachePolicy;
  let testHeaders, testCookies, testQueryStrings;

  beforeEach(async () => {
    testHeaders = ["header1"];
    testCookies = ["cookie1"];
    testQueryStrings = ["queryString1"];
    testCachePolicy = {
      minTTL: 10,
      name: "cachePolicyOne",
      comment: "comment",
      defaultTTL: 10,
      maxTTL: 100,
      parametersInCacheKeyAndForwardedToOrigin: {
        cookiesConfig: {
          cookieBehavior: "whitelist",
          cookies: testCookies
        },
        headersConfig: {
          headerBehavior: "whitelist",
          headers: testHeaders
        },
        queryStringsConfig: {
          queryStringBehavior: "whitelist",
          queryStrings: testQueryStrings
        },
        enableAcceptEncodingGzip: true,
        enableAcceptEncodingBrotli: true
      }
    };

    mockCreateDistributionPromise.mockResolvedValueOnce({
      Distribution: {
        Id: "distribution123"
      }
    });
    mockCreateCachePolicyPromise.mockResolvedValueOnce({
      CachePolicy: {
        Id: "TestCachePolicyId"
      }
    });

    component = await createComponent();

    await component.default({
      origins: [
        {
          url: "https://exampleorigin.com",
          pathPatterns: {
            "/some/path": {
              cachePolicy: testCachePolicy
            }
          }
        }
      ]
    });
  });

  it("creates the cache policy before the behavior", () => {
    expect(mockCreateCachePolicy.mock.calls[0][0]).toMatchSnapshot();
  });

  it("attaches cache policy to cache behavior", async () => {
    assertHasCacheBehavior(mockCreateDistribution, {
      PathPattern: "/some/path",
      CachePolicyId: "TestCachePolicyId"
    });

    expect(mockCreateDistribution.mock.calls[0][0]).toMatchSnapshot();
  });
});
