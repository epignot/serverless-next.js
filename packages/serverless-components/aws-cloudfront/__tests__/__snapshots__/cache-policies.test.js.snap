// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cache Policies attaches cache policy to cache behavior 1`] = `
Object {
  "DistributionConfig": Object {
    "Aliases": Object {
      "Items": Array [],
      "Quantity": 0,
    },
    "CacheBehaviors": Object {
      "Items": Array [
        Object {
          "AllowedMethods": Object {
            "CachedMethods": Object {
              "Items": Array [
                "GET",
                "HEAD",
              ],
              "Quantity": 2,
            },
            "Items": Array [
              "GET",
              "HEAD",
            ],
            "Quantity": 2,
          },
          "CachePolicyId": "TestCachePolicyId",
          "Compress": true,
          "DefaultTTL": undefined,
          "FieldLevelEncryptionId": "",
          "ForwardedValues": Object {
            "Cookies": Object {
              "Forward": "all",
            },
            "Headers": Object {
              "Items": Array [],
              "Quantity": 0,
            },
            "QueryString": true,
            "QueryStringCacheKeys": Object {
              "Items": Array [],
              "Quantity": 0,
            },
          },
          "LambdaFunctionAssociations": Object {
            "Items": Array [],
            "Quantity": 0,
          },
          "MaxTTL": undefined,
          "MinTTL": undefined,
          "PathPattern": "/some/path",
          "SmoothStreaming": false,
          "TargetOriginId": "exampleorigin.com",
          "TrustedSigners": Object {
            "Enabled": false,
            "Quantity": 0,
          },
          "ViewerProtocolPolicy": "https-only",
        },
      ],
      "Quantity": 1,
    },
    "CallerReference": "1566599541192",
    "Comment": "",
    "CustomErrorResponses": Object {
      "Items": Array [],
      "Quantity": 0,
    },
    "DefaultCacheBehavior": Object {
      "AllowedMethods": Object {
        "CachedMethods": Object {
          "Items": Array [
            "HEAD",
            "GET",
          ],
          "Quantity": 2,
        },
        "Items": Array [
          "HEAD",
          "GET",
        ],
        "Quantity": 2,
      },
      "Compress": false,
      "DefaultTTL": 86400,
      "FieldLevelEncryptionId": "",
      "ForwardedValues": Object {
        "Cookies": Object {
          "Forward": "none",
        },
        "Headers": Object {
          "Items": Array [],
          "Quantity": 0,
        },
        "QueryString": false,
        "QueryStringCacheKeys": Object {
          "Items": Array [],
          "Quantity": 0,
        },
      },
      "LambdaFunctionAssociations": Object {
        "Items": Array [],
        "Quantity": 0,
      },
      "MaxTTL": 31536000,
      "MinTTL": 0,
      "SmoothStreaming": false,
      "TargetOriginId": "exampleorigin.com",
      "TrustedSigners": Object {
        "Enabled": false,
        "Items": Array [],
        "Quantity": 0,
      },
      "ViewerProtocolPolicy": "redirect-to-https",
    },
    "Enabled": true,
    "HttpVersion": "http2",
    "Origins": Object {
      "Items": Array [
        Object {
          "CustomHeaders": Object {
            "Items": Array [],
            "Quantity": 0,
          },
          "CustomOriginConfig": Object {
            "HTTPPort": 80,
            "HTTPSPort": 443,
            "OriginKeepaliveTimeout": 5,
            "OriginProtocolPolicy": "https-only",
            "OriginReadTimeout": 30,
            "OriginSslProtocols": Object {
              "Items": Array [
                "TLSv1.2",
              ],
              "Quantity": 1,
            },
          },
          "DomainName": "exampleorigin.com",
          "Id": "exampleorigin.com",
          "OriginPath": "",
        },
      ],
      "Quantity": 1,
    },
    "PriceClass": "PriceClass_All",
  },
}
`;

exports[`Cache Policies creates the cache policy before the behavior 1`] = `
Object {
  "CachePolicyConfig": Object {
    "Comment": "comment",
    "DefaultTTL": 10,
    "MaxTTL": 100,
    "MinTTL": 10,
    "Name": "cachePolicyOne",
    "ParametersInCacheKeyAndForwardedToOrigin": Object {
      "CookiesConfig": Object {
        "CookieBehavior": "whitelist",
        "Cookies": Object {
          "Items": Array [
            "cookie1",
          ],
          "Quantity": 1,
        },
      },
      "EnableAcceptEncodingBrotli": undefined,
      "EnableAcceptEncodingGzip": undefined,
      "HeadersConfig": Object {
        "HeaderBehavior": "whitelist",
        "Headers": Object {
          "Items": Array [
            "header1",
          ],
          "Quantity": 1,
        },
      },
      "QueryStringsConfig": Object {
        "QueryStringBehavior": "whitelist",
        "QueryStrings": Object {
          "Items": Array [
            "queryString1",
          ],
          "Quantity": 1,
        },
      },
    },
  },
}
`;
